glib-macros/src/gflags_attribute.rs

```diff
@@ -0,0 +1,184 @@
+    }
+}
+
+// Generate gobject_sys::GFlagsValue structs mapping the enum such as:
+//     gobject_sys::GFlagsValue {
+//         value: MyFlags::A.bits(),
+//         value_name: "The Name\0" as *const _ as *const _,
+//         value_nick: "nick\0" as *const _ as *const _,
+//     },
+fn gen_gflags_values(
+    enum_name: &Ident,
+    enum_variants: &Punctuated<Variant, Comma>,
+) -> (TokenStream, usize) {
+    // start at one as GFlagsValue array is null-terminated
+    let mut n = 1;
+    let recurse = enum_variants.iter().filter(|v| { !attribute_has_skip(&v.attrs) } ).map(|v| {
+        let name = &v.ident;
+        let mut value_name = name.to_string().to_camel_case();
+        let mut value_nick = name.to_string().to_kebab_case();
+
+        let attrs = parse_item_attributes("gflags", &v.attrs);
+        let attrs = match attrs {
+            Ok(attrs) => attrs,
+            Err(e) => abort_call_site!(
+                "{}: gflags enum supports only the following optional attributes: #[gflags(name = \"The Name\", nick = \"the-nick\")] or #[gflags(skip)]",
+                e
+            ),
+        };
+
+        attrs.into_iter().for_each(|attr|
+            match attr {
+                ItemAttribute::Name(n) => value_name = n,
+                ItemAttribute::Nick(n) => value_nick = n,
+            }
+        );
+
+        let value_name = format!("{}\0", value_name);
+        let value_nick = format!("{}\0", value_nick);
+
+        n += 1;
+        quote_spanned! {v.span()=>
+            gobject_sys::GFlagsValue {
+                value: #enum_name::#name.bits(),
+                value_name: #value_name as *const _ as *const _,
+                value_nick: #value_nick as *const _ as *const _,
+            },
+        }
+    });
+    (
+        quote! {
+            #(#recurse)*
+        },
+        n,
+    )
+}
+
+fn gen_bitflags(enum_name: &Ident, enum_variants: &Punctuated<Variant, Comma>) -> TokenStream {
+    let recurse = enum_variants.iter().map(|v| {
+        let name = &v.ident;
+        let disc = v.discriminant.as_ref().expect("missing discriminant");
+        let value = &disc.1;
+
+        quote_spanned! {v.span()=>
+            const #name = #value;
+        }
+    });
+
+    quote! {
+        bitflags::bitflags! {
+            struct #enum_name: u32 {
+                #(#recurse)*
+            }
+        }
+    }
+}
+
+pub fn impl_gflags(input: &DeriveInput, gtype_name: &LitStr) -> TokenStream {
+    let name = &input.ident;
+
+    let enum_variants = match input.data {
+        Data::Enum(ref e) => &e.variants,
+        _ => abort_call_site!("gflags only supports enums"),
+    };
+
+    let bitflags = gen_bitflags(name, enum_variants);
+
+    let get_type = format_ident!("{}_get_type", name.to_string().to_snake_case());
+    let (gflags_values, nb_gflags_values) = gen_gflags_values(name, enum_variants);
+
+    quote! {
+        #bitflags
+
+        impl glib::translate::ToGlib for #name {
```

Comment:
Won't the `glib::` part in the generated code fail if you add a test using the macro to the `glib` crate?

Reply:
Should be replace with `$crate`, indeed.

Reply:
That's not how it works with proc macros, but yes :)

Reply:
For proc-macros, I already provided a solution in the other PR of @gdesmott :)

Reply:
Right, we forgot to fix the other macro which I used as a base for this one. Fixed both.